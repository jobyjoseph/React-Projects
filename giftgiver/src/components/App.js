import React from "react";
import { Button } from "react-bootstrap";
import Gift from "./Gift";

class App extends React.Component {
  state = {
    gifts: [],
  };

  addGift = () => {
    const { gifts } = this.state;
    const ids = gifts.map((ele) => ele.id);
    const max = Math.max(...ids, 0);
    gifts.push({ id: max + 1 });
    this.setState({ gifts });
  };

  removeGift = (id) => {
    const { gifts } = this.state;
    const newGifts = gifts.filter((gift) => gift.id !== id);
    this.setState((prevState) => {
      return {
        ...prevState,
        gifts: newGifts,
      };
    });
  };

  render() {
    return (
      <div>
        <h1>Gift Giver</h1>
        <div className="gift-list">
          {this.state.gifts.map((gift) => (
            <Gift key={gift.id} gift={gift} removeGift={this.removeGift} />
          ))}
        </div>
        <Button className="btn-add" onClick={this.addGift}>
          Add Gift
        </Button>
      </div>
    );
  }
}

export default App;
