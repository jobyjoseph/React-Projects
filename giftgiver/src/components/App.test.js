import React from "react";
import { shallow } from "enzyme";
import App from "./App";

describe("App", () => {
  const app = shallow(<App />);

  it("renders correctly", () => {
    expect(app).toMatchSnapshot(); // check if a component renders correctly
  });

  it("initializes the `state` with an empty list of gifts", () => {
    expect(app.state().gifts).toEqual([]);
  });

  describe("when clicking the `add-gift` button", () => {
    const id = 1;

    beforeEach(() => {
      app.find(".btn-add").simulate("click");
    });

    afterEach(() => {
      app.setState({ gifts: [] });
    });

    it("adds a new gift to `state`", () => {
      expect(app.state().gifts).toEqual([{ id }]);
    });

    it("renders a new gift list item", () => {
      expect(app.find(".gift-list").children().length).toEqual(1);
    });

    it("renders `Gift` component", () => {
      expect(app.find("Gift").exists()).toBe(true);
    });

    describe("and remove the gift on remove button click", () => {
      beforeEach(() => {
        app.instance().removeGift(id);
      });

      it("removes gift from `state`", () => {
        expect(app.state().gifts).toEqual([]);
      });
    });
  });
});
