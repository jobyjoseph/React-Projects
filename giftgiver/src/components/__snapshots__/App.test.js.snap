// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`App renders correctly 1`] = `
<div>
  <h1>
    Gift Giver
  </h1>
  <div
    className="gift-list"
  />
  <Button
    active={false}
    className="btn-add"
    disabled={false}
    onClick={[Function]}
    variant="primary"
  >
    Add Gift
  </Button>
</div>
`;
